### ISSUE TITLE
[7.5.0] large torrent stuck on building circuits.

### ISSUE THREAD

**Issue by Dmole** [Tribler/tribler#5399]
"Filesize 269.2 GB in 0 files"
100% CPU for 15+ minutes on this 22MB torrent.
Can't remove the torrent, had to kill -9 Tribler and restore a backup of the ~/.Tribler folder

**Comment by ichorid** [comment]
Unfortunately, there is almost nothing we can do without more information.
@Dmole , could you provide us the torrent file that caused the issue? Or a glitchy `.Tribler` folder?

**Comment by Dmole** [comment]
magnet:?xt=urn:btih:cccaad8d8b856c9609137890d5802189d841ad2c

**Comment by ichorid** [comment]
Confirmed, trying to download this one results in "Stopped on error"

**Comment by devos50** [comment]
@ichorid do you want to debug this one?

**Comment by synctext** [comment]
"Filesize 269.2 GB in 0 files"
This already provides 90% certainty what is going on. Probably a .torrent file with numerous MBytes of filenames and hashes. We should test Tribler with a 10+ MByte legal torrent file. Thnx @qstokkink for finding this! Indeed one of the reasons people will dump Tribler forever.

**Comment by qstokkink** [comment]
Since I complained about this behavior, I guess I'll assign myself to solve it. 😃

**Comment by qstokkink** [comment]
I found a single torrent that managed to lock up my Tribler, on an AMD Ryzen Threadripper 3970X 😃. After some investigation, the culprit - for me at least - seems to be the REST API's `Request: GET http://localhost:[xxx]/downloads?get_pieces=1` locking up the main thread for almost 5 seconds. Roughly 4/5 seconds were spent in this loop:

https://github.com/Tribler/tribler/blob/633f8bdda3bdca26a2ea96208ee489c9312f157a/src/tribler/core/components/libtorrent/restapi/downloads_endpoint.py#L146-L155

As a POC, I applied some quick-and-dirty optimizations to get the 5 seconds down to ~1 millisecond and the locking-up behavior stopped.

**Comment by qstokkink** [comment]
Annoying, tables with tens of thousands of rows slow down Qt _a lot_. So, even though files are streamed in through the REST API with my quick-and-dirty fix, the user interface becomes slow after some time anyway if you click on the "Files" tab.

The only sane solution is to implement pagination for a torrent's files but this will require a substantial refactoring effort (that I was initially trying to avoid).

**Comment by qstokkink** [comment]
As the second part of this issue, I figured out the slowdown on adding torrents. Our current logic is as follows when adding a torrent file:

1. Read the torrent file ourselves.
2. Bdecode the file contents ourselves.
3. Construct the metainfo dictionary ourselves.
4. Feed the metainfo into libtorrent.

This is a lot of processing on the main thread for big torrents. We might gain a lot of responsiveness by diverting this processing from the main thread for big torrent files. Alternatively, we could leave the file loading to libtorrent: I'll look through the git history to try and find why we don't do this already.

**Comment by kozlovsky** [comment]
> We might gain a lot of responsiveness by diverting this processing from the main thread for big torrent files.

I agree that offloading the processing to a separate thread could significantly enhance responsiveness for large torrent files and provide a smoother user experience

**Comment by qstokkink** [comment]
I'll split the fix for this issue up into three PRs (in no particular order):

- [ ] Stop requesting all pieces and all files for each download in the downloads page and only request the details for the download that has its details showing.
- [ ] Leave torrent loading to libtorrent.
- [ ] Refactor the torrent details file tab and make it paginated.

**Comment by qstokkink** [comment]
The underlying data structure (tree) to provide fast pagination of a torrent's (directory and file) structure [is turning out to be a sizable implementation (click me for POC terminal-based torrent explorer)](https://github.com/qstokkink/tribler/blob/large_torrent_support/src/tribler/core/components/libtorrent/torrent_loading_test.py). For logical separation of PRs, I'll split that off into a fourth PR:

- [ ] Create a tree data structure to view torrent file structures in a paginated fashion.

After this has been merged, it can be hooked into the downloads endpoint to serve the REST API.

**Comment by qstokkink** [comment]
The "final integration" PR is nearing 1k lines. I'll split off PR 5/4 (🙂) with just the non-REST changes to keep things small enough to be reviewable.

**Comment by qstokkink** [comment]
I'm in the process of writing the last unit tests and docstrings for the new file list pagination. After #7694 has been merged, I'll drop the "final" PR. When that has also been merged, our code will _still_ not be as optimized as possible nor will it be as pretty as it can be. However, I consider those optimizations "features" and I'll leave further improvements for future work.
